File: Week_02/144/binary_tree_preorder_traversal.py
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        res = []

        def pre_order(node: TreeNode):
            if node is not None:
                res.append(node.val)
                pre_order(node.left)
                pre_order(node.right)

        pre_order(root)
        return res

    def test(self):
        n1 = TreeNode(1)
        n2 = TreeNode(2)
        n3 = TreeNode(3)

        n1.right = n2
        n2.left = n3

        ans = self.preorderTraversal(n1)
        assert ans == [1, 2, 3], f"ans: {ans}"
        print("well done")


if __name__ == "__main__":
    Solution().test()

File: Week_02/144/binary_tree_preorder_traversal.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func preorderTraversal(root *TreeNode) []int {
	res := make([]int, 0)

	var preOrder func(node *TreeNode)
	preOrder = func(node *TreeNode) {
		if node != nil {
			res = append(res, node.Val)
			preOrder(node.Left)
			preOrder(node.Right)
		}
	}
	preOrder(root)
	return res
}
